### ISSUE TITLE
Single small object detection - Dataset guidelines?

### ISSUE THREAD

**Issue by aagrawal357** [ultralytics/ultralytics#4166]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi,

I am building an object detection model on a custom dataset. I have only one object, say banana. Objects in the training dataset are usually small (<10-20% of the image size).  I have followed the following guidelines for my dataset distribution:

<img width="988" alt="image" src="https://github.com/ultralytics/ultralytics/assets/93498440/712ff7a4-a752-4c0b-9140-72a1f6ccdc8e">

My current training dataset has ~15k images (~15k object instances) and has ~2.2k background images (about 15%). Labels are accurate, consistent and verified.  Images have variety to include different lighting, weather conditions etc..

I am using the large model (with >43M parameters), starting from pre-trained weights and am training all the layers. Results are not satisfactory at all. Sometimes, the model detects a car as a banana with very high confidence (>0.6). Sometimes, it fails to identify a clearly visible banana. What would you recommend in this case?

One thing I can tell you is that my background images don't have any cars. Can that be the reason for a car false positive? If yes, are there any guidelines around what to include in the background images? Let's say I include cars and car false positives go down, how do I know which other objects to include as background and how many images for each of those objects?

I would really appreciate your help with this. Thank you!

**Comment by glenn-jocher** [comment]
@aagrawal357 hello,

Thanks for reaching out and for already exploring the documentation and best practices extensively.

Detecting smaller objects can indeed be a challenge, especially when, as in your case, they take up only a small part of the image. Here are some suggestions:

1. **Anchor Boxes:** In YOLOv8, Anchor Boxes play a key role. These are predefined rectangular dimensions that the model expects the objects to exhibit. You might want to verify if your anchor boxes are a good match for the dimensions of your "banana" objects. Adjusting these to better suit your specific objects may improve your results.
    
2. **Batch & Subdivision Sizes:** Choosing an appropriate batch and subdivision size can have a noticeable effect on the detection accuracy for smaller objects. You might want to experiment with different sizes.
    
3. **Resolution:** The input resolution of images can have a significant impact on detecting smaller objects. High-resolution images could be beneficial for detecting smaller objects better.
    
4. **Background Images:** Including negatives (images that do not contain your object of interest) in your dataset is a good practice. However, the lack of cars in your background images should not cause the model to false-detect cars as bananas, as object detection models learn to detect the presence of the objects you're interested in, rather than the absence of unannotated objects. That being said, maintaining a diverse set of background images that represent the types of environments where you expect to deploy your model can be beneficial.
    
Ultimately, you may not need to worry too much about which objects to include in the background. The model is designed to generalize as much as possible, so the goal is to include a variety of backgrounds and scenarios. If you notice consistent false positives with a specific object (ex. it continually mistakes cars for bananas), adding some images with those objects could help, although it's important to remember that these objects should not be annotated as bananas, since they are not the objects of interest.

Lastly, remember that machine learning involves a lot of experimentation. You may have to tweak and tune your parameters and data to achieve satisfactory results. Keep trying and good luck with your model!

**Comment by Mercury829** [comment]
Hello, may I know in which folder are the Anchor Boxes for YOLOv8 located?

**Comment by glenn-jocher** [comment]
@Mercury829 hello,

The anchor boxes for YOLOv8 are not located in a specific folder but rather they are derived directly during the model's training process. They are computed on the dataset during training to best fit the distribution of object shapes and sizes. This computation is performed by the k-means clustering algorithm over the width and height of the bounding boxes in your dataset.

However, if you wish to view or modify the initial anchors used, they can be found in the .yaml configuration file for the corresponding model. You'll look for a field named 'anchors' which is followed by a series of values. These values represent the initial anchor boxes used by the model.

Thank you for your interest in YOLOv8 and please let us know if you have any further questions.

**Comment by glenn-jocher** [comment]
@wcyjerry i apologize for any confusion. You're correct, YOLOv8 does not rely on preset anchor boxes as previous versions did. Instead, it considers every grid cell in the image as a potential anchor point, which allows for better localization of objects of different sizes and aspect ratios. It does not not utilize k-means clustering to determine anchor boxes. Thank you for pointing out the error, and I'm sorry for any misunderstanding. Your understanding of YOLOv8 is correct. Let me know if you have any other questions!
